### PR TITLE
Proposal for Exposing Storage Metrics

### PR DESCRIPTION
**Pull request by jingxu97**
This proposal is exploring ways of exposing storage metrics to users

### CODE REVIEW

**Comment by jingxu97** [comment]
cc @kubernetes/sig-instrumentation-feature-requests @kubernetes/sig-storage-feature-requests

File: contributors/design-proposals/storage-metrics.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+```
+### Register the VolumeStats metrics to Prometheus 
+
+The following metrics could be registered to Prometheus
+
+
+| Metric name | Metric type | Labels/tags |
+|-------------|-------------|-------------|
+| volume_stats_capacityBytes | Gauge | namespace=\<persistentvolumeclaim-namespace\> <br/> persistentvolumeclaim=\<persistentvolumeclaim-name\>  <br/> persistentvolume=\<persistentvolume-name\> |
```
Comment (brancz) [inline contributors/design-proposals/storage-metrics.md]: mixed snake and camel case

File: contributors/design-proposals/storage-metrics.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+	PVRef PVReference `json:"podRef"`
+	// Embedded FsStats
+	FsStats
+	// Name is the name given to the Volume
+	// +optional
+	Name string `json:"name,omitempty"`
+}
+
+// PVCReference contains enough information to locate the referenced PVC.
+type PodReference struct {
```
Comment (brancz) [inline contributors/design-proposals/storage-metrics.md]: Comment says PVCReference, struct name is PodReference

File: contributors/design-proposals/storage-metrics.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	Name      string `json:"name"`
+	Namespace string `json:"namespace"`
+	UID       string `json:"uid"`
+}
+
+// PVReference contains enough information to locate the referenced PV.
+type PodReference struct {
```
Comment (brancz) [inline contributors/design-proposals/storage-metrics.md]: comment says PVReference, struct name is PodReference

File: contributors/design-proposals/storage-metrics.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+ - Resource Metrics API (Alpha feature in progress [Resource Metrics API](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/resource-metrics-api.md)
+ - Collect metrics in Heapster. (metrics has to exposed by Kubelet in Summary API)
+ - Export metrics through StackDriver 
+ - Kubectl Command (kubectl top, kubectl describe)
+
+All the storage metrics mentioned in Background are collected by Kubelet Summary API so that it could be integrated with the above monitoring system. However, pod volume metrics is provided in PodStat object’s VolumeStat and VolumeStat only has volume name specified in pod specification. There is no direct way of checking volume usage information by PVC name, which is more preferable by users.
+
+
+## Goals
+The gola of this proposal is to expose storage usage metrics to users for monitoring their storage systems.
```
Comment (zhangxiaoyu-zidif) [inline contributors/design-proposals/storage-metrics.md]: Is this feature on pod-level ?

Comment (jingxu97) [reply contributors/design-proposals/storage-metrics.md]: I think the metric is per volume. Could you explain what you mean by "pod-level"?

**Comment by piosz** [comment]
Is this ready for review? I still can see `[WIP]` tag in the first line of the doc.

File: contributors/design-proposals/storage-metrics.md (hunk 4 above)
Comment (surajssd) [reply contributors/design-proposals/storage-metrics.md]: spell: `s/gola/goal/`

File: contributors/design-proposals/storage-metrics.md [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+## Proposal
+
+This proposal has three parts, the first part is trying to address the issue of adding volume usage information indexed by PVC and PV name in Kubelet Summary API. The second part is to register the volume metrics to . The third part is to add more storage metrics to Heapster.
+
+
+### volume usage information indexed by PVC/PV in Kubelet Summary API
+
+The basic idea is to cache PVC and the volume information in kubelet volume manager which is similar to caching the pod and volume information. In Summary API, In Summary API, we could add PVC and PV references into VolumeStats which is included in PodStats.
```
Comment (derekwaynecarr) [inline contributors/design-proposals/storage-metrics.md]: nit: double "In Summary API"

Comment (kmova) [reply contributors/design-proposals/storage-metrics.md]: nit: incomplete sentence? "The second part is to register the volume metrics to ."

**Comment by MrD2** [comment]
Status?
Was it done?

**Comment by nodesocket** [comment]
👍 I'd like to view how much of a persistent volume is in use and don't know of a way.

**Pull request closed without merging**